WooCommerce/Classes/ViewRelated/Orders/Order Details/Shipping Labels/Create Shipping Label Form/Package Details/ShippingLabelPackageDetailsViewModel.swift

```diff
@@ -93,11 +98,46 @@ final class ShippingLabelPackageDetailsViewModel: ObservableObject {
         self.weightUnit = weightUnit
         self.packagesResponse = packagesResponse
         self.selectedPackageID = selectedPackageID
+        self.initialTotalWeight = totalWeight
+
+        configureItemRows()
+        configureTotalWeights()
         configureResultsControllers()
         syncProducts()
         syncProductVariations()
         setDefaultPackage()
-        setTotalWeight(using: totalWeight)
+    }
+
+    /// Observe changes in products and variations to update item rows.
+    ///
+    private func configureItemRows() {
+        $products.combineLatest($productVariations) { [weak self] (products, variations) in
+            guard let self = self else { return [] }
+            return self.generateItemsRows(products: products, productVariations: variations)
+        }
+        .assign(to: &$itemsRows)
+    }
+
+    /// Observe changes in selected custom package, products and variations to update total package weight.
+    ///
+    private func configureTotalWeights() {
+        $selectedCustomPackage.combineLatest($products, $productVariations)
+            .filter { [weak self] _ in self?.isPackageWeightEdited == false }
```

Comment:
It felt a bit strange to use an "external" variable as the filter. If I'm understanding the code correctly, when `initialTotalWeight` is set, that's the weight that will be used regardless of any UI updates (I don't see this getting any new values when `isPackageWeightEdited` is `true`, and I see no code setting it back to `false`).

If that's the case, maybe we skip the combine logic when there is an initial value. I haven't tested it, but something like:
https://gist.github.com/koke/375587ea5c3a80c2ea756bbc51265580

Reply:
Thank you @koke for taking a look! 

The reason why I keep `initialTotalWeight` inside the data stream is because we have to calculate the total weight before comparing it with the initial value. The calculation depends on values selected package, product and variation list, so comparing the initial value with the total weight directly is not what I intend to do.

`isPackageWeightEdited` is set to `true` in 2 cases:
- when user enters any arbitrary value for the weight (this is updated when text field is edited)
- when the initial value is different from the calculated weight.

Since We intend to not calculate package weight after we detect that user has updated the weight manually, `isPackageWeightEdited` is never set to false. This is also the reason why I use this variable to filter the stream. It would not make sense to make it a publisher because we only want to use it for the condition check. There has to be a better way for this but I'm not sure how.

Reply:
@koke  I have updated `configureTotalWeight` with `initialTotalWeight` as input and use it in a separate stream in 30b50093a90dfa27b3f515ef8cb387a3dd1f65db - please let me know what you think about this. (I was distracted so the commit message doesn't convey all the idea)

Reply:
If I'm understanding this correctly, you want to peek at the first value emitted, and if it doesn't match an existing initial value, then we don't want to update the values anymore. I think this implementation should work, but using `share` to handle the two cases and the `isPackageWeightEdited` to flag the case where we stop processing values seems a bit off.

I'd like to suggest leveraging publisher completion for that case, which might make for something that's a bit easier to read. I don't think there's a `completeIf` operator, but we can fake that with `tryMap`/`catch`. 

I'm starting to feel sick from the vaccine so I might have missed something obvious, but here's a totally untested idea:

```swift
private func configureTotalWeights(initialTotalWeight: String?) {
    self.totalWeight = initialTotalWeight ?? ""

    $selectedCustomPackage.combineLatest($products, $productVariations)
        .map { [weak self] (customPackage, products, variations) -> Double in
            self?.calculateTotalWeight(products: products, productVariations: variations, customPackage: customPackage) ?? 0
        }
        .tryMap { numericWeight in
            let calculatedWeight = String(numericWeight)

            // Set the total weight as the initial value if it was input manually;
            // otherwise use the calculated weight.
            if let initialTotalWeight = initialTotalWeight,
               calculatedWeight != initialTotalWeight {
                throw Errors.overridenTotalWeight
            }
            return calculatedWeight
        }
        .catch({ _ in
            Empty()
        })
        .assign(to: &$totalWeight)
}
```

With the error defined:

```swift
private extension ShippingLabelPackageDetailsViewModel {
    enum Errors: Error {
        case overridenTotalWeight
    }
}
```

If I understand correctly how Combine works, sending a completion should prevent `assign` from storing any new values. According to the docs:

> The Subscribers.Assign instance created by this operator maintains a strong reference to object, and sets it to nil when the upstream publisher completes (either normally or with an error).

Reply:
Hi @koke - I like the idea of using error to stop the stream! However my intention is to check with `initialTotalWeight` only after the first calculation. Keeping that check in the middle of the stream causes failure when we have a new calculated weight that is different from the initial value. Separating into 2 streams seem to be safer I think.

As for `isPackageWeightEdited`, since this value is also updated from the view (when user start entering the weight manually), I don't think we should omit it and rely only on `initialTotalWeight`. Instead, in 627e49fc9bf5619964fe78bd6c89f9d8064feaf6 I have made `isPackageWeightEdited` a publisher and observe it to throw an error when needed. WDYT?

Reply:
Sorry for the delay, I'm not sure why but I didn't get a notification for this comment.

> However my intention is to check with initialTotalWeight only after the first calculation

I looked into this more closely, and I think I finally understand the intention now. I hadn't realized the total package weight was editable in the app (the field didn't seem editable to me).

It seems to me that we don't save explicitly whether the user manually edited the weight, so we have to rely on inspecting whether the saved weight matches the selected package. I understand you only want to check the first time it calculates because selecting a different package could give a different weight result.

Now that I think I understand this better, I think separating it in two streams works and it's the right logic, although I'd go back to my first comment. While testing this, I see we configure the results controllers before calling `configureTotalWeights`, so all the values we need for the first weight calculation are already present on initialization, which means we don't have to wait asynchronously for that to be published. It seems `assign(to:)` tries to immediately get a value, and if you set a breakpoint on the `map` that compares the initial value, you can see that the  `init` method is still in the call stack.

Maybe this initial check doesn't really need combine and can happen synchronously, while we still use combine to update the value (as long as `isPackageWeightEdited` is `false`). Maybe this reads a bit more easily, since it's clearer that the initial value comparison isn't really async:

```swift
    private func configureTotalWeights(initialTotalWeight: String?) {
        let calculatedWeight = String(calculateTotalWeight(products: products, productVariations: productVariations, customPackage: selectedCustomPackage))
        if let initialTotalWeight = initialTotalWeight,
           initialTotalWeight != calculatedWeight {
            isPackageWeightEdited = true
            // No need to observe changes if weight was edited
            return
        }

        $selectedCustomPackage.combineLatest($products, $productVariations)
            .map { [weak self] (customPackage, products, variations) -> Double in
                self?.calculateTotalWeight(products: products, productVariations: variations, customPackage: customPackage) ?? 0
            }
            .combineLatest($isPackageWeightEdited)
            .tryMap { (weight, isEdited) in
                guard !isEdited else {
                    throw Errors.overridenTotalWeight
                }
                return String(weight)
            }
            .catch { _ in Empty() }
            .assign(to: &$totalWeight)
    }
```

On top of that, I thought of a slightly nicer approach to completing the stream when `isPackageWeightEdited` changes, although I haven't tested it. Instead of `tryMap`, we can use `prefix(while:)` on the combined stream, and stop when `isPackageWeightEdited` is true:

```swift
    private func configureTotalWeights(initialTotalWeight: String?) {
       // ...Initial checks...

        $selectedCustomPackage.combineLatest($products, $productVariations)
            .map { [weak self] (customPackage, products, variations) -> Double in
                self?.calculateTotalWeight(products: products, productVariations: variations, customPackage: customPackage) ?? 0
            }
            .map { String($0) }
            .combineLatest($isPackageWeightEdited)
            .prefix(while: { _, isPackageWeightEdited in !isPackageWeightEdited })
            .map { weight, _ in weight }
            .assign(to: &$totalWeight)
    }
```

I also found a new issue while testing this. It seems like, according to the docs, the `onEditingChanged` of `TextField` will be called _"when the user begins editing text and after the user finishes editing text"_, which is when the UI sets `isPackageWeightEdited` as `true`, even if the value hasn't changed. So as soon as you tap on the editable field, it's flagged as manually edited

Reply:
Also, for reference, it looks like the web actually [persists in its state](https://github.com/Automattic/woocommerce-services/blob/dbeed610f336f8db7995743765dbcd0ff2e18474/client/extensions/woocommerce/woocommerce-services/state/shipping-label/reducer.js#L355) when the weight has been manually changed, so that might be an even simpler alternative: persist `isPackageWeightEdited` in the selected package entity.